### PR TITLE
Add osx-arm64 build

### PR DESCRIPTION
**Pull request by vespakoen**
Checklist
* [x] Used a [personal fork of the feedstock to propose changes](https://conda-forge.org/docs/maintainer/updating_pkgs.html#forking-and-pull-requests)
* [x] Bumped the build number (if the version is unchanged)
* [ ] Reset the build number to `0` (if the version changed)
* [x] [Re-rendered]( https://conda-forge.org/docs/maintainer/updating_pkgs.html#rerendering-feedstocks ) with the latest `conda-smithy` (Use the phrase <code>@<space/>conda-forge-admin, please rerender</code> in a comment in this PR for automated rerendering)
* [ ] Ensured the license file is being packaged.

### CODE REVIEW

**Comment by bkpoon** [comment]
You probably need to keep the `vtk` dependency. The error seems to be that `cmake` is looking for a file in the `site-packages` directory, specifically a .so file.
```
2023-01-07T17:20:21.6453010Z CMake Error at /Users/runner/miniforge3/conda-bld/ocp_1673111572484/_h_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_place/lib/cmake/vtk-9.2/VTKPython-targets.cmake:440 (message):
2023-01-07T17:20:21.6454590Z   The imported target "VTK::vtkCommonCorePython" references the file
2023-01-07T17:20:21.6455000Z 
2023-01-07T17:20:21.6456790Z      "/Users/runner/miniforge3/conda-bld/ocp_1673111572484/_h_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_place/lib/python3.11/site-packages/vtkmodules/vtkCommonCore.so"
2023-01-07T17:20:21.6457750Z 
2023-01-07T17:20:21.6458210Z   but this file does not exist.  Possible reasons include:
2023-01-07T17:20:21.6458500Z 
2023-01-07T17:20:21.6458940Z   * The file was deleted, renamed, or moved to another location.
2023-01-07T17:20:21.6459220Z 
2023-01-07T17:20:21.6459660Z   * An install or uninstall procedure did not complete successfully.
2023-01-07T17:20:21.6459990Z 
2023-01-07T17:20:21.6460410Z   * The installation package was faulty and contained
2023-01-07T17:20:21.6460720Z 
2023-01-07T17:20:21.6462300Z      "/Users/runner/miniforge3/conda-bld/ocp_1673111572484/_h_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_place/lib/cmake/vtk-9.2/VTKPython-targets.cmake"
2023-01-07T17:20:21.6463410Z 
2023-01-07T17:20:21.6463890Z   but not all the files it references.
```
For cross-compiling, the `site-packages` directory is deleted (https://github.com/conda-forge/cross-python-feedstock/blob/main/recipe/activate-cross-python.sh) so `cmake` will not be able to find that file. Do you know if this package is linking against something inside `site-packages`? If not, you can probably patch out that check for `osx-arm64` so that `cmake` can continue.

Also, the `python` dependency should have the `# [build_platform != target_platform]` selector (https://conda-forge.org/docs/maintainer/knowledge_base.html#cross-compilation).

**Comment by vespakoen** [comment]
Thanks for the help, especially the link to activate-cross-python.sh, I am new to conda and not sure where some things are happening / coming from.

I **think** the main issue I have is that the vtk dependency that is being installed is the `osx-x86_64` one, and not the `osx-arm64` one, I downloaded both of them and in one of them the file names to the .so files in vtkmodules have a different extension.

**Comment by vespakoen** [comment]
I added your suggestions but it is stuck at the same point.
In the logs it looks like `vtk @ 9.2.2-qt_py311h4d5bf31_204` is installed, when I search for that version identifier over here: https://anaconda.org/conda-forge/vtk/files it points to the osx-64 version.

Since the cross compiling is using some hacks because it doesn't run on a native arm64 machine, perhaps this is expected? 

Here is what I am finding inside both manually downloaded packages:

# osx-arm64

Screenshot 2023-01-09 at 01 10 16

VTKPython-targets-release.cmake
```cmake
# Import target "VTK::vtkCommonCorePython" for configuration "Release"
set_property(TARGET VTK::vtkCommonCorePython APPEND PROPERTY IMPORTED_CONFIGURATIONS RELEASE)
set_target_properties(VTK::vtkCommonCorePython PROPERTIES
  IMPORTED_COMMON_LANGUAGE_RUNTIME_RELEASE ""
  IMPORTED_LOCATION_RELEASE "${_IMPORT_PREFIX}/lib/python3.11/site-packages/vtkmodules/vtkCommonCore.so"
  IMPORTED_NO_SONAME_RELEASE "TRUE"
  )
```

# osx-64

Screenshot 2023-01-09 at 01 03 41

VTKPython-targets-release.cmake:

```cmake
# Import target "VTK::vtkCommonCorePython" for configuration "Release"
set_property(TARGET VTK::vtkCommonCorePython APPEND PROPERTY IMPORTED_CONFIGURATIONS RELEASE)
set_target_properties(VTK::vtkCommonCorePython PROPERTIES
  IMPORTED_COMMON_LANGUAGE_RUNTIME_RELEASE ""
  IMPORTED_LOCATION_RELEASE "${_IMPORT_PREFIX}/lib/python3.11/site-packages/vtkmodules/vtkCommonCore.cpython-311-darwin.so"
  IMPORTED_NO_SONAME_RELEASE "TRUE"
  )
```

So, perhaps the `VTKPython-targets-release.cmake` is loaded from the `osx-arm64` package, but the osx-64 package is the one that is installed?

I'm out of ideas ;)

**Comment by whophil** [comment]
@vespakoen `vtk` should be a host dependency, but you appear to have changed it to a build dependency.

**Comment by whophil** [comment]
@vespakoen Cross-compiling issues aside, I'm not sure that the OCP library can be successfully compiled on Mac ARM machines. Have you reviewed these threads?

- https://github.com/CadQuery/OCP/issues/68
- https://github.com/CadQuery/cadquery/issues/891

To further complicate this, this recipe is using pre-generated binding code from the upstream project. This binding code varies from platform to platform. If the binding code generated on Mac ARM differs from the binding code generated on Mac x86 machines, that would require that the upstream project release a set of Mac ARM bindings as well. I think this would require Azure Pipelines to support Mac ARM machines (see https://github.com/microsoft/azure-pipelines-agent/issues/3184), or for somebody to manually generate these bindings on their Mac ARM machine and upload the release artifacts with each release.

**Comment by adam-urbanczyk** [comment]
> @vespakoen Cross-compiling issues aside, I'm not sure that the OCP library can be successfully compiled on Mac ARM machines. Have you reviewed these threads?
> 
> * [Compiling OCP for Apple M1 Max (arm64) CadQuery/OCP#68](https://github.com/CadQuery/OCP/issues/68)
> * [Installing cadquery on macOS on Apple silicon CadQuery/cadquery#891](https://github.com/CadQuery/cadquery/issues/891)
> 
> To further complicate this, this recipe is using pre-generated binding code from the upstream project. This binding code varies from platform to platform. If the binding code generated on Mac ARM differs from the binding code generated on Mac x86 machines, that would require that the upstream project release a set of Mac ARM bindings as well. I think this would require Azure Pipelines to support Mac ARM machines (see [microsoft/azure-pipelines-agent#3184](https://github.com/microsoft/azure-pipelines-agent/issues/3184)), or for somebody to manually generate these bindings on their Mac ARM machine and upload the release artifacts with each release.

It *probably* does not depend on the hw, just the os.

**Comment by bkpoon** [comment]
From https://github.com/CadQuery/OCP/issues/68#issuecomment-1045118232, it does look like it should compile and things should mostly run?

I am not familiar with how this package wraps code, so I cannot comment on whether or not the bindings need to be regenerated with `pywrap`.

For the `cmake` error, the check will need to be patched out. Otherwise, the build will never get past that check. Alternatively, if you do need to link against that file in `site-packages` during the build, you will need to create a separate conda environment with the `osx-arm64` version of the package so that the file exists and then tell `cmake` how to find that file. Something similar is done with some packages that depend on `pytorch` because the `torch` libraries are also in `site-packages`. An example is in`openmm-torch-feedstock` in the `rc` branch (https://github.com/conda-forge/openmm-torch-feedstock/blob/c5d938303f0d40a108b3739d6303182f65a2b75b/recipe/build.sh#L7-L15).

The difference in the library name between `osx-64` and `osx-arm64` should be okay as long as it is consistent. You should not try to link against `osx-64` libraries in your `osx-arm64` build.

**Comment by adam-urbanczyk** [comment]
> From [CadQuery/OCP#68 (comment)](https://github.com/CadQuery/OCP/issues/68#issuecomment-1045118232), it does look like it should compile and things should mostly run?
> 
> I am not familiar with how this package wraps code, so I cannot comment on whether or not the bindings need to be regenerated with `pywrap`.
> 
> For the `cmake` error, the check will need to be patched out. Otherwise, the build will never get past that check. Alternatively, if you do need to link against that file in `site-packages` during the build, you will need to create a separate conda environment with the `osx-arm64` version of the package so that the file exists and then tell `cmake` how to find that file. Something similar is done with some packages that depend on `pytorch` because the `torch` libraries are also in `site-packages`. An example is in`openmm-torch-feedstock` in the `rc` branch (https://github.com/conda-forge/openmm-torch-feedstock/blob/c5d938303f0d40a108b3739d6303182f65a2b75b/recipe/build.sh#L7-L15).
> 
> The difference in the library name between `osx-64` and `osx-arm64` should be okay as long as it is consistent. You should not try to link against `osx-64` libraries in your `osx-arm64` build.

I don't know full background of this PR, but it looks like some builds are simply failing due to cmake not finding OCCT (opencascade).  Nothing to do with site-packages AFAICT. If you know the dir layout when cross-compiling, just specify the location of OCCT when calling cmake.

**Comment by whophil** [comment]
Hi @looooo, I think you solved this issue for cross-compiling FreeCAD for OSX ARM, do you remember what you did to make it work?

```
  The imported target "VTK::vtkCommonCorePython" references the file

     "/Users/runner/miniforge3/conda-bld/ocp_1673255405836/_h_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_place/lib/python3.11/site-packages/vtkmodules/vtkCommonCore.so"

  but this file does not exist.  Possible reasons include:
```

**Comment by vespakoen** [comment]
The last build worked, using the trick described by @bkpoon over here: https://github.com/conda-forge/openmm-torch-feedstock/blob/c5d938303f0d40a108b3739d6303182f65a2b75b/recipe/build.sh#L7-L15

I am still wondering why adding `vtk` to the `host` dependencies didn't work.
I still believe I was getting the `osx-64` version of the vtk package instead of the `osx-arm64` one.

I am not sure and I would like to find out.

To test this out locally, I installed the x86_64 version of miniforge, and ran:
```shell
conda build recipe -m .ci_support/osx_arm64_.yaml
```

And changed the CONDA_BUILD_SYSROOT in `osx_arm64_.yaml` to the result of `xcrun --show-sdk-path`. (there is probably a better way to change the build sysroot (perhaps with SDKROOT?), but this worked for me ;)

Anyways thanks for all the help, let me know if there is anything that has to change to get this merged.

File: recipe/meta.yaml [hunk 1]
```diff
@@ -1,6 +1,7 @@
 {% set name = "ocp" %}
 {% set version = "7.6.3.0" %}
 {% set occt_ver = "7.6.3" %}
+{% set py_version = "3.11" %}
```
Comment (whophil) [inline recipe/meta.yaml:4]: This is currently only building for Python 3.11, we would want to open this up to have the complete build matrix again.

File: recipe/build.sh [hunk 2]
```diff
@@ -1,11 +1,26 @@
...
+fi
+
 # FindOpenCascade.cmake bundled with OCP references the env var `CONDA_PREFIX`.
 # That is the right prefix when manually running CMake inside a conda env, but
 # the wrong one when using conda-build. Substitute the right prefix inline.
-CONDA_PREFIX="${PREFIX}" cmake -B build -S "${SRC_DIR}/src" \
+CONDA_PREFIX="${PREFIX}" cmake ${CMAKE_ARGS} -B build -S "${SRC_DIR}/src" \
 	-G Ninja \
+	-DCMAKE_PREFIX_PATH="${VTK_DIR}" \
+	-DPython3_FIND_STRATEGY=LOCATION \
```
Comment (whophil) [inline recipe/build.sh:21]: I wonder if these `Python3*` flags are required. Did you try without?

The CMakeLists file being processed here is [this one](https://github.com/CadQuery/pywrap/blob/d3bc7321dab7d82340b4fe16defb2578cf6610de/bindgen/CMakeLists.j2), which uses `find_package(pybind11 REQUIRED)` **without**  calling `find_package(python)` before it.

I don't know which Python hint locations are used by this command.

https://pybind11.readthedocs.io/en/stable/cmake/index.html

File: .ci_support/osx_arm64_.yaml [hunk 3]
```diff
@@ -9,20 +9,19 @@ cxx_compiler:
...
 - '4'
-python:
-- 3.9.* *_cpython
 target_platform:
-- osx-64
+- osx-arm64
 vtk:
 - 9.2.2
+
+CONDA_BUILD_SYSROOT:
```
Comment (whophil) [inline .ci_support/osx_arm64_.yaml:25]: I have never modified this file directly, do you know if this change persists a re-rendering of the feedstock?

**Review by whophil — CHANGES_REQUESTED** [review]

**Comment by vespakoen** [comment]
Thanks for the suggestions, I pushed it over to this other PR: https://github.com/conda-forge/ocp-feedstock/pull/20

Over there, I got rid of the hack, or, let's say I used a smaller hack to find vtk from the build requirements ;)

File: recipe/build.sh (hunk 2 above)
Comment (vespakoen) [reply recipe/build.sh:21]: I tried to remove the FindPython stuff but it fails to build on Linux then, so I added them back ;)

File: recipe/meta.yaml (hunk 1 above)
Comment (vespakoen) [reply recipe/meta.yaml:4]: Done

**Comment by vespakoen** [comment]
The VTK module is the one using FindPython3 by the way, so that is what we need the `Python3_XXX` variables for.

**Comment by bkpoon** [comment]
Great! One additional thing is to delete the environment on `osx-arm64`. Otherwise, it will get packaged as well. You can also try putting it in a different location. The example code is a little further down (https://github.com/conda-forge/openmm-torch-feedstock/blob/c5d938303f0d40a108b3739d6303182f65a2b75b/recipe/build.sh#L52-L55)

You can also test the packages by storing the build as Azure artifacts and then manually installing the built packages. The artifacts can be stored by adding
```
azure:
  store_build_artifacts: true
```
to your `conda-forge.yml` (https://conda-forge.org/docs/maintainer/conda_forge_yml.html#azure). Unzip the artifact and there should be a directory for the platform. The package will be in that directory. The `osx-arm64` packages should be roughly the same size as the `osx-64` packages. I strongly suggest that the package be tested on an Apple Silicon machine to verify that is works as expected.

> I am still wondering why adding vtk to the host dependencies didn't work.
I still believe I was getting the osx-64 version of the vtk package instead of the osx-arm64 one.
>
> I am not sure and I would like to find out.

The `build` section if mainly for build tools on the building platform and the `host` is for dependencies on the target platform (https://docs.conda.io/projects/conda-build/en/stable/resources/define-metadata.html#requirements-section). For cross-compiling, the `build` sections will be for x86-64 and the `host` sections will be Apple Silicon for `osx-arm64`. So adding dependencies to the `host` section generally will work. However, for `vtk`, there is a library in `site-packages` that is checked by `cmake`. The cross-compiling tool will delete everything in `site-packages` so the check for that library will always fail during cross-compiling. If that library were in `${PREFIX}/lib`, then this hack would not be necessary.

**Comment by whophil** [comment]
Great insights, thanks @bkpoon . Do you happen to recall what exactly VTK is looking for inside `site-packages`?

**Comment by vespakoen** [comment]
@bkpoon thanks for those insights! I wonder if in stead of removing it afterwards, choosing a different path (for example /tmp?) would be a better option, in the example that you gave and which I copied, the arm64-osx deps are installed inside of the build/recipe folder, which is an odd location to begin with?

Over at https://github.com/conda-forge/ocp-feedstock/pull/20 I got rid of the "install arm64-osx deps" thingy in the build.sh, and instead added `-DCMAKE_PREFIX_PATH="${BUILD_PREFIX}"` to the CMAKE_ARGS, which also worked (see: https://dev.azure.com/conda-forge/feedstock-builds/_build/results?buildId=640604&view=logs&jobId=15f8dc28-9f6f-5e3a-5b3c-2b169071e5be). I guess I could make that more specific by using something like VTK_DIR=BUILD_PREFIX` though.

@whophil VTK was looking for `lib/python3.11/site-packages/vtkmodules/vtkCommonCore.so`

**Comment by bkpoon** [comment]
@vespakoen Yes! You should be able to use /tmp or some other writeable directory on the Azure vm. I think I just originally picked `${RECIPE_DIR}` on a whim and since it worked, did not bother with trying anything else.

Also, for this PR and for #20, you have `occt` in both the `build` and `host` sections. You should remove the one in `build` because during cross-compilation, you will get the `osx-64` and maybe the `osx-arm64` versions in `${PREFIX}/lib`. In the build logs for "Also add back FindPython3 for Windows" in #20, you get warning like
```
2023-01-12T19:17:02.5791920Z ld: warning: ignoring file $BUILD_PREFIX/lib/libvtkCommonCore-9.2.9.2.2.dylib, building for macOS-arm64 but attempting to link with file built for macOS-x86_64
```
(https://dev.azure.com/conda-forge/feedstock-builds/_build/results?buildId=640604&view=results).
That seems to indicate that the build does not have the correct platform for linking. (And it looks like this library is from `occt`, so maybe the files from `vtk` are not linked to anything in this build).

**Comment by bkpoon** [comment]
The new path is not being used by `cmake`. I'm trying to debug in #21.

**Comment by bkpoon** [comment]
So I got it to work with these 2 changes

1. Add `python` to the creation of the conda environment for `vtk`. Otherwise, the version might not match the one for that specific build.
2. Copy `site-packages` back to `${PREFIX}`. I could not figure out how to get `cmake` to recognize the `vtk` environment so copying to the original location satisfied the `cmake` check.

The successful build is in #21 under the "Seems like copying only works" commit (https://github.com/conda-forge/ocp-feedstock/pull/21/commits/4c0853024009f628ea2d5061a6901cf193cc3799).

**Comment by adam-urbanczyk** [comment]
Hey, could you explain what is actually happening here? What is the nature of the problem? Is it a VTK packaging bug? Is it a bug in the provided target definitions? Does the artifact that is needed actually exist in the VTK package? If so and all fails, you can always define the vtk-related targets yourself.

**Comment by bkpoon** [comment]
It's specific to how `osx-arm64` is cross-compiled for `conda-forge`. The `site-packages` directory with the dependencies from `host` is deleted by the `cross-python` dependency. This causes the check from `cmake` for a file inside the `site-packages` directory to fail.

The workaround is to point `cmake` to a `site-packages` directory that has the file or in this case to copy the files back into `site-packages`.

**Comment by adam-urbanczyk** [comment]
Thanks @bkpoon, then I'd say it is a `cross-python` bug?

I think the best workaround would be to define a custom `FindVTK.cmake` that defines the actually needed targets
```
VTK::WrappingPythonCore VTK::RenderingCore VTK::CommonDataModel VTK::CommonExecutionModel
```

**Comment by bkpoon** [comment]
I don't think it's a bug. I'm not familiar with the design process, but I suspect this is a way of handling situations where Python is needed on the build platform to get some code for building (or is needed for `conda-build` to run). The target platform version cannot be used because of the difference in architecture. That's why `meta.yaml` has these duplicate dependencies just for cross-compiling. In one case that I have come across (`pyside2`), the build actually calls some Python code to generate something for the build. That requires that dependency to be in the `build` section. So instead of trying to manage both architectures in the same `${PREFIX}` directory, the files for the target architecture are removed.

Yes, another workaround is to patch out the check in `cmake`. As long as the required components do not trigger some check for files in `site-packages`, the build should work without the hack.

**Pull request merged**